### PR TITLE
Fixed issue BTS-624

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Backport of https://github.com/arangodb/arangodb/pull/14974/

Enterprise companion PR: https://github.com/arangodb/enterprise/pull/814

Bug-fix related to the ticket https://arangodb.atlassian.net/browse/BTS-624.
The `move-calculations-up` optimization rule is now also applied to subqueries, when they don't have dependencies on the outer nodes, don't have modification nodes and don't read their own writes. This fixed the execution of a query without the splicing-subqueries option being faster than the execution of a query with this option (after version 3.8, this option cannot be switched off). 

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] Backport for 3.8: https://github.com/arangodb/arangodb/pull/15013
- [x] Backport for 3.7: https://github.com/arangodb/arangodb/pull/15019

#### Related Information

- [x] Docs PR: https://github.com/arangodb/docs/pull/795
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/814
- [x] GitHub issue / Jira ticket number: https://arangodb.atlassian.net/browse/BTS-624
- [ ] Design document: 

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [x] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

File: arangod/Aql/OptimizerRules.cpp [hunk 1]
```diff
@@ -1719,13 +1720,28 @@ void arangodb::aql::moveCalculationsUpRule(Optimizer* opt,
...
+      }
+    } else {
+      // if it's a subquery node, it cannot move upwards if there's a
+      // modification keyword in the subquery e.g.
+      // INSERT would not be scope-limited by the outermost subqueries, so we could end up
+      // inserting a smaller amount of documents than what's actually proposed in the query.
+      auto nn = ExecutionNode::castTo<SubqueryNode*>(n);
+      if (nn->isModificationNode()) { 
+        continue;
+      }
```
Comment (mpoeter) [inline arangod/Aql/OptimizerRules.cpp:1743]: Is this not already covered by the previous `isDeterministic` check?

**Review by mpoeter — APPROVED** [review]

File: arangod/Aql/OptimizerRules.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Aql/OptimizerRules.cpp:1743]: Yes! It won't do any harm leaving the check in here, but it is redundant. 
As it is a nice-to-have cleanup, shall we clean it up in all backports, or only do it in devel? I would prefer the latter.

Comment (mpoeter) [reply arangod/Aql/OptimizerRules.cpp:1743]: I am ok with only doing it in devel.

Comment (jsteemann) [reply arangod/Aql/OptimizerRules.cpp:1743]: Thanks!

**Pull request merged**